bolt-modules/boltlib/lib/puppet/functions/wait.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+# Wait for a Future or array of Futures to finish and return results.
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:wait, Puppet::Functions::InternalFunction) do
+  # Wait for a Future or array of Futures to finish and return results.
+  # @param futures A Bolt Future object or array of Bolt Futures to wait on.
+  # @param timeout How long to wait for Futures to finish before raising a Timeout error.
+  # @option options [Boolean] _catch_errors Whether to catch raised errors.
+  # @return A Result or Results from the Futures
+  # @example Upload a large file in the background, then wait until it's loaded
+  #   $futures = background() || {
+  #     upload_file("./very_large_file", "/opt/jfrog/artifactory/var/etc/artifactory", $targets)
+  #   }
+  #   # Run an unrelated task
+  #   run_task("deploy", $targets)
+  #   # Wait for the file upload to finish
+  #   $results = wait($futures)
+  dispatch :wait do
+    param 'Variant[Future, Array[Future]]', :futures
+    # optional_param does not set Integer to optional
+    optional_param 'Optional[Numeric]', :timeout
+    optional_param 'Hash[String[1], Any]', :options
+    return_type 'Array[Boltlib::PlanResult]'
+  end
```

Comment:
I think this needs separate dispatches since both `timeout` and `options` are optional parameters. The following function invocation:

```
wait($future, '_catch_errors' => true)
```

Results in the following error:

```
"'wait' parameter 'timeout' expects a value of type Undef or Numeric, got Struct"
```

Reply:
Oh, nice catch! Definitely needs two dispatches.